File: apps/ios/WindmillKit/Tests/WindmillGymTests/PickerSixRankingTests.swift
import XCTest
@testable import WindmillGym

// C2 · what "the six most-used" means on this surface. They are ranked from the log THIS DEVICE
// holds, over its newest fifty sessions, and topped up in order from the shared opener list — so a
// fresh account still opens on six, and an account with a log opens on its own six. The twin of
// Android's `PickerOptions.mostTrained` and of web's `trainedCounts`.
final class PickerSixRankingTests: XCTestCase {

    // The catalog order is the tie-break, so the fixture states it once here.
    private let catalog = [
        Exercise(id: "back-squat", name: "Back Squat"),
        Exercise(id: "bench-press", name: "Bench Press"),
        Exercise(id: "deadlift", name: "Deadlift"),
        Exercise(id: "overhead-press", name: "Overhead Press"),
        Exercise(id: "barbell-row", name: "Barbell Row"),
        Exercise(id: "chin-up", name: "Chin Up"),
        Exercise(id: "leg-press", name: "Leg Press"),
        Exercise(id: "lat-pulldown", name: "Lat Pulldown"),
        Exercise(id: "face-pull", name: "Face Pull"),
    ]

    private func session(_ named: [String], at day: Int64 = 0) -> SessionSummary {
        SessionSummary(session: Session(id: "s_\(day)_\(named.joined(separator: "-"))",
                                        startedAtMs: day * 86_400_000,
                                        finishedAtMs: day * 86_400_000 + 3_600_000),
                       setCount: named.count, exercises: named)
    }

    private func six(_ sessions: [SessionSummary], taken: [String] = []) -> [String] {
        PickerOptions.matching(query: "", catalog: catalog, taken: taken, sessions: sessions)
            .six.map(\.id)
    }

    func testALogLessAccountOpensOnTheOpenersInTheirOwnOrder() {
        XCTAssertEqual(PickerOptions.openers,
                       ["back-squat", "bench-press", "deadlift", "overhead-press", "barbell-row",
                        "chin-up"])
        XCTAssertEqual(six([]), PickerOptions.openers,
                       "a fresh account has nothing to rank and must still see six")
    }

    // The count is SESSIONS THAT NAMED IT, never working sets: nothing on the wire ranks a movement
    // by use, and this invents no read.
    func testTheSixAreTheMovementsTheMostSessionsNamed() {
        let log = [
            session(["Face Pull", "Lat Pulldown"], at: 3),
            session(["Face Pull", "Leg Press"], at: 2),
            session(["Face Pull", "Lat Pulldown"], at: 1),
        ]
        XCTAssertEqual(six(log),
                       ["face-pull", "lat-pulldown", "leg-press", "back-squat", "bench-press",
                        "deadlift"],
                       "three sessions named the face pull, two the pulldown, one the leg press")
    }

    // A session the LOG served names its movements by name; one this device wrote names them by id
    // (`LocalLog.summaries()`). One movement, either spelling, one count.
    func testAMovementIsCountedUnderEitherSpellingOfItself() {
        let mixed = [session(["face-pull"], at: 2), session(["Face Pull"], at: 1)]
        XCTAssertEqual(six(mixed).first, "face-pull")
        XCTAssertEqual(Array(six([session(["Leg Press"], at: 3),
                                  session(["Leg Press"], at: 2),
                                  session(["face-pull"], at: 1)]).prefix(2)),
                       ["leg-press", "face-pull"],
                       "one movement under two spellings outranked a movement named twice")
    }

    // Naming the same movement twice in one session is one session, not two.
    func testASessionCountsOnceHoweverManyTimesItNamesAMovement() {
        let twice = [session(["Face Pull", "Face Pull", "Face Pull"], at: 2),
                     session(["Leg Press"], at: 1),
                     session(["Leg Press"], at: 0)]
        XCTAssertEqual(Array(six(twice).prefix(2)), ["leg-press", "face-pull"])
    }

    func testMovementsTrainedEquallyOftenKeepCatalogOrder() {
        let level = [session(["Face Pull", "Lat Pulldown", "Leg Press"], at: 1)]
        XCTAssertEqual(six(level),
                       ["leg-press", "lat-pulldown", "face-pull", "back-squat", "bench-press",
                        "deadlift"],
                       "the tie is broken by the catalog and not by the order the log named them")
    }

    // A fixed depth, so a phone that has paged further back does not rank differently from one that
    // has not.
    func testOnlyTheNewestFiftySessionsAreCounted() {
        XCTAssertEqual(PickerOptions.trainedWindow, 50)
        let newest = (0..<50).map { session(["Lat Pulldown"], at: Int64(100 - $0)) }
        let older = (0..<20).map { session(["Face Pull"], at: Int64(50 - $0)) }
        XCTAssertEqual(six(newest + older),
                       ["lat-pulldown", "back-squat", "bench-press", "deadlift", "overhead-press",
                        "barbell-row"],
                       "a session below the window ranked a movement anyway")
    }

    // The top-up never repeats a movement the ranking already found, and it never lengthens the six.
    func testAShortRankingIsToppedUpFromTheOpenersWithoutRepeatingOne() {
        let ranked = six([session(["Deadlift", "Face Pull"], at: 1)])
        XCTAssertEqual(ranked,
                       ["deadlift", "face-pull", "back-squat", "bench-press", "overhead-press",
                        "barbell-row"])
        XCTAssertEqual(Set(ranked).count, 6, "the top-up drew a movement the ranking already had")
    }

    // Everything already in this session is gone from the catalog before the ranking runs.
    func testAMovementAlreadyInTheSessionIsNeverOneOfTheSix() {
        let taken = ["face-pull", "back-squat"]
        let ranked = six([session(["Face Pull", "Back Squat", "Leg Press"], at: 1)], taken: taken)
        XCTAssertEqual(ranked,
                       ["leg-press", "bench-press", "deadlift", "overhead-press", "barbell-row",
                        "chin-up"])
    }

    // A typed query features nothing at all: the six are the empty query's answer.
    func testATypedQueryStillFeaturesNothing() {
        let typed = PickerOptions.matching(query: "pull", catalog: catalog, taken: [],
                                           sessions: [session(["Face Pull"], at: 1)])
        XCTAssertTrue(typed.six.isEmpty)
        XCTAssertEqual(typed.matches.map(\.id), ["lat-pulldown", "face-pull"])
    }

    // A catalog that never landed has nothing to rank, and says so in the bytes all three surfaces say.
    func testAnUnreadCatalogRanksNothingAndKeepsItsOwnSentence() {
        let silent = PickerOptions.matching(query: "", catalog: [], taken: [],
                                            sessions: [session(["Face Pull"], at: 1)])
        XCTAssertTrue(silent.six.isEmpty)
        XCTAssertEqual(silent.empty, "The catalog didn’t load. It comes back when you have signal.")
    }
}

File: apps/android/gym/src/test/kotlin/works/windmill/gym/ui/PickerSixRankingTests.kt
package works.windmill.gym.ui

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import works.windmill.gym.domain.Exercise
import works.windmill.gym.domain.SessionSummary
import works.windmill.gym.domain.TheSix

// C2: an empty query shows six most-used movements and then the WHOLE catalogue, on every empty
// query — never gated on a first session. The six are ranked off the last fifty sessions this device
// holds and topped up in order from the shared opener list, so a fresh account still sees six.
class PickerSixRankingTests {
    private val openers = TheSix.movements
    private val extras = (1..12).map { Exercise(id = "ex_$it", name = "Movement $it") }
    private val catalog = openers + extras

    private fun session(id: Int, vararg named: String) = SessionSummary(
        id = "ses_$id",
        startedAtMs = id.toLong(),
        finishedAtMs = id.toLong(),
        exercises = named.toList(),
    )

    @Test
    fun testAnEmptyQueryShowsTheSixAndThenTheWholeCatalogueUncapped() {
        val options = PickerOptions.matching(query = "", catalog = catalog, taken = emptyList())

        assertEquals(6, options.six.size)
        assertEquals("the rest of the catalogue follows, uncapped",
                     catalog.size - 6, options.matches.size)
        assertTrue("and nothing is listed twice",
                   options.matches.none { row -> options.six.any { it.id == row.id } })
    }

    @Test
    fun testTheSixAreRankedFromTheLogAndToppedUpFromTheOpeners() {
        val sessions = listOf(
            session(3, "Movement 5", "Movement 9"),
            session(2, "Movement 5"),
            session(1, "Movement 9"),
        )

        val options = PickerOptions.matching(query = "", catalog = catalog, taken = emptyList(),
                                             sessions = sessions)

        assertEquals(
            listOf("Movement 5", "Movement 9", "Back Squat", "Bench Press", "Deadlift", "Overhead Press"),
            options.six.map { it.name },
        )
    }

    // The device's own sessions name their movements by ID; the log's name them by NAME. A movement is
    // counted by either spelling of itself, or a shelf full of training would rank nothing.
    @Test
    fun testAShelvedSessionCountsTheSameAsOneTheLogServed() {
        val sessions = listOf(session(2, "ex_7"), session(1, "ex_7"), session(3, "Movement 4"))

        val options = PickerOptions.matching(query = "", catalog = catalog, taken = emptyList(),
                                             sessions = sessions)

        assertEquals(listOf("Movement 7", "Movement 4"), options.six.take(2).map { it.name })
    }

    @Test
    fun testTheRankingReadsFiftySessionsAndNoMore() {
        val recent = (1..50).map { session(it, "Movement 1") }
        val older = (51..80).map { session(it, "Movement 2") }

        val options = PickerOptions.matching(query = "", catalog = catalog, taken = emptyList(),
                                             sessions = recent + older)

        assertEquals("Movement 1", options.six.first().name)
        assertTrue("the fifty-first session and beyond rank nothing",
                   options.six.drop(1).none { it.name == "Movement 2" })
    }

    @Test
    fun testAFreshAccountStillSeesSixAndTheyAreTheOpenersInOrder() {
        val options = PickerOptions.matching(query = "", catalog = catalog, taken = emptyList(),
                                             sessions = emptyList())

        assertEquals(
            listOf("Back Squat", "Bench Press", "Deadlift", "Overhead Press", "Barbell Row", "Chin Up"),
            options.six.map { it.name },
        )
    }

    @Test
    fun testATypedQueryFeaturesNothingAndKeepsTheSevenRowCap() {
        val many = (1..20).map { Exercise(id = "press_$it", name = "Press $it") }

        val options = PickerOptions.matching(query = "press", catalog = many, taken = emptyList(),
                                             sessions = listOf(session(1, "Press 3")))

        assertTrue(options.six.isEmpty())
        assertEquals(PickerOptions.shown, options.matches.size)
    }

    @Test
    fun testAMovementAlreadyInTheSessionIsNotOfferedAsOneOfTheSix() {
        val options = PickerOptions.matching(query = "", catalog = catalog,
                                             taken = listOf("back-squat", "bench-press"),
                                             sessions = emptyList())

        assertEquals(
            listOf("Deadlift", "Overhead Press", "Barbell Row", "Chin Up"),
            options.six.take(4).map { it.name },
        )
        assertTrue(options.six.none { it.id == "back-squat" || it.id == "bench-press" })
    }

    // C11: the same bytes web and iOS say.
    @Test
    fun testTheCatalogueThatDidNotLoadSaysWhatTheOtherSurfacesSay() {
        val options = PickerOptions.matching(query = "bench", catalog = emptyList(), taken = emptyList())

        assertEquals("The catalog didn’t load. It comes back when you have signal.", options.unread)
        assertNull("one silence, one sentence", options.empty)
        assertNull(options.create)
    }
}
